backend/queues/listingCreatedProcessor.js

```diff
@@ -0,0 +1,181 @@
+const ethers = require('ethers')
+
+const { marketplaceAbi } = require('@origin/utils/marketplace')
+
+const { Network, Shop, Transaction } = require('../models')
+const queues = require('./queues')
+const { ListingID } = require('../utils/id')
+const { getIpfsHashFromBytes32 } = require('../utils/_ipfs')
+const { getLogger } = require('../utils/logger')
+const { IS_TEST } = require('../utils/const')
+const { Sentry } = require('../sentry')
+const { TransactionStatuses } = require('../enums')
+
+const log = getLogger('listingCreatedProcessor')
+
+// Wait for 2 blocks confirmation before considering a tx mined.
+const NUM_BLOCKS_CONFIRMATION = IS_TEST ? 0 : 2
+
+/**
+ * Function to start the queue processing.
+ */
+function attachToQueue() {
+  const queue = queues['listingCreatedQueue']
+  queue.process(processor)
+  queue.resume() // Start if paused
+}
+
+/**
+ * Waits for a blockchain ListingCreated transaction to get confirmed,
+ * then updates the associate shop's ListingId.
+ *
+ * @param {Object} job: Bull job object.
+ * job.data is expected to have the following fields:
+ *   {string} txHash: hash of the ListingCreated tx to watch.
+ *   {string} fromAddress: Address that sent the transaction
+ *   {number} shopId: Id of the shop
+ * @returns {Promise<null||{receipt: ethers.TransactionReceipt, listingId: number, offerId: number, ipfsHash: string}>}
+ * @throws
+ */
+async function processor(job) {
+  const queueLog = (progress, str) => {
+    job.log(str)
+    job.progress(progress)
+  }
+  const jobId = `${job.queue.name}-${job.id}` // Prefix with queue name since job ids are not unique across queues.
+
+  const { txHash, fromAddress, shopId } = job.data
+  log.info(
+    `Waiting for ListingCreated tx. ShopId: {shopId} Hash: ${txHash} fromAddress: ${fromAddress} `
+  )
+  let confirmation
+
+  try {
+    // Load the associated shop.
+    const shop = await Shop.findOne({ where: { id: shopId } })
+    if (!shop) {
+      throw new Error(`Failed loading shop with id ${shopId}`)
+    }
+
+    const network = await Network.findOne({
+      where: { networkId: shop.networkId, active: true }
+    })
+    const provider = new ethers.providers.JsonRpcProvider(network.provider)
+    const marketplace = new ethers.Contract(
+      network.marketplaceContract,
+      marketplaceAbi,
+      provider
+    )
+
+    // Load the transaction from the DB.
+    const transaction = await Transaction.findOne({
+      networkId: network.id,
+      shopId,
+      fromAddress,
+      hash: txHash
+    })
+    if (!transaction) {
+      throw new Error(
+        `No transaction found in the DB with hash ${txHash} for shop ${shopId} and wallet ${fromAddress}`
+      )
+    }
+
+    // Load the tx from the blockchain based on its hash.
+    queueLog(25, 'Loading tx from the blockchain')
+    const tx = await provider.getTransaction(txHash)
```

Comment:
If it's in the mempool but hasn't been mined yet, this would throw? Clients will get their txHash and send to the backend before it's mined.

Reply:
I think this returns the transaction that was sent to be mined. So it should not error out if the transaction is still in the mempool. I will double check though.

Reply:
Yea I was just thinking that if the user is using MetaMask with whatever provider, that TX might not necessarily be available in the Alchemy mempool by the time this code runs

Reply:
Gotcha. Yes that's definitely a possibility. Let me either add a retry inline or will rely on the queue retry to handle this case.